x/auth/vesting/types/vesting_account.go

```diff
@@ -611,3 +615,297 @@ func marshalYaml(i interface{}) (interface{}, error) {
+}
+
+// TrackDelegation tracks a desired delegation amount by setting the appropriate
+// values for the amount of delegated vesting, delegated free, and reducing the
+// overall amount of base coins.
+func (va *ClawbackVestingAccount) TrackDelegation(blockTime time.Time, balance, amount sdk.Coins) {
+	va.BaseVestingAccount.TrackDelegation(balance, va.GetVestingCoins(blockTime), amount)
+}
+
+// GetStartTime returns the time when vesting starts for a periodic vesting
+// account.
+func (va ClawbackVestingAccount) GetStartTime() int64 {
+	return va.StartTime
+}
+
+// GetVestingPeriods returns vesting periods associated with periodic vesting account.
+func (va ClawbackVestingAccount) GetVestingPeriods() Periods {
+	return va.VestingPeriods
+}
+
+// coinEq returns whether two Coins are equal.
+// The IsEqual() method can panic.
+func CoinEq(a, b sdk.Coins) bool {
+	return a.IsAllLTE(b) && b.IsAllLTE(a)
+}
+
+// Validate checks for errors on the account fields
+func (va ClawbackVestingAccount) Validate() error {
+	if va.GetStartTime() >= va.GetEndTime() {
+		return errors.New("vesting start-time must be before end-time")
+
+	}
+
+	lockupEnd := va.GetStartTime()
+	lockupCoins := sdk.NewCoins()
+
+	for _, p := range va.LockupPeriods {
+		lockupEnd += p.Length
+		lockupCoins = lockupCoins.Add(p.Amount...)
+	}
+
+	if lockupEnd > va.EndTime {
+		return errors.New("lockup schedule extends beyond account end time")
+	}
+
+	// use coinEq to prevent panic
+	if !CoinEq(lockupCoins, va.OriginalVesting) {
+		return errors.New("original vesting coins does not match the sum of all coins in lockup periods")
+	}
+
+	vestingEnd := va.GetStartTime()
+	vestingCoins := sdk.NewCoins()
+
+	for _, p := range va.VestingPeriods {
+		vestingEnd += p.Length
+		vestingCoins = vestingCoins.Add(p.Amount...)
+	}
+
+	if vestingEnd > va.EndTime {
+		return errors.New("vesting schedule exteds beyond account end time")
+	}
+
+	if !CoinEq(vestingCoins, va.OriginalVesting) {
+		return errors.New("original vesting coins does not match the sum of all coins in vesting periods")
+	}
+
+	return va.BaseVestingAccount.Validate()
+}
+
+type clawbackGrantAction struct {
+	funderAddress       string
+	grantStartTime      int64
+	grantLockupPeriods  []Period
+	grantVestingPeriods []Period
+	grantCoins          sdk.Coins
+}
+
+func NewClawbackGrantAction(
+	funderAddress string,
+	grantStartTime int64,
+	grantLockupPeriods, grantVestingPeriods []Period,
+	grantCoins sdk.Coins) exported.AddGrantAction {
+	return clawbackGrantAction{
+		funderAddress:       funderAddress,
+		grantStartTime:      grantStartTime,
+		grantLockupPeriods:  grantLockupPeriods,
+		grantVestingPeriods: grantVestingPeriods,
+		grantCoins:          grantCoins,
+	}
+}
+
+func (cga clawbackGrantAction) AddToAccount(ctx sdk.Context, rawAccount exported.VestingAccount) error {
+	cva, ok := rawAccount.(*ClawbackVestingAccount)
+	if !ok {
+		return fmt.Errorf("expected *ClawbackVestingAccount, got %T", rawAccount)
```

Comment:
In case the name changes in the future
```suggestion
		return fmt.Errorf("expected %T, got %T", &ClawbackVestingAccount{}, rawAccount)
```